src/server/middleware/auth_checker.rs

```diff
@@ -0,0 +1,197 @@
+use std::str::FromStr;
+
+use actix_http::httpmessage::HttpMessage;
+use actix_web::{dev::ServiceRequest, dev::ServiceResponse, http, http::StatusCode, Error};
+
+use crate::{
+    models::auth as auth_model,
+    server::error::{ErrorCode, MatrixError},
+};
+
+use std::task::{Context, Poll};
+
+use actix_service::{Service, Transform};
+use futures::future::{ok, FutureExt, LocalBoxFuture, Ready};
+
+pub struct AuthChecker;
+
+impl AuthChecker {
+    pub fn new() -> Self {
+        AuthChecker {}
+    }
+}
+
+impl<S, B> Transform<S> for AuthChecker
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+    B: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type InitError = ();
+    type Transform = AuthCheckerMiddleware<S>;
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(AuthCheckerMiddleware { service })
+    }
+}
+
+pub struct AuthCheckerMiddleware<S> {
+    service: S,
+}
+
+fn get_token_from_query(query_string: &str) -> Option<&str> {
```

Comment:
Actix does not provide any query string parser?

Reply:
No, it doesn't, it provided before ver 1.0, but then they removed it. Also in Middleware it doesn't seem to be convenient to use extractors, so decided to write the stuff manually because it is simple enough.

Reply:
just fyi, serde_urlencoded is for this purpose. this way you don't have to worry about escapes.